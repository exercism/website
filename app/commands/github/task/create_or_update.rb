module Github
  class Task
    class CreateOrUpdate
      include Mandate

      initialize_with :issue_url, :attributes

      def call
        task = ::Github::Task.create_or_find_by!(issue_url:) do |t|
          t.attributes = attributes
        end

        task.update!(attributes)
        task
      end
    end
  end
end
