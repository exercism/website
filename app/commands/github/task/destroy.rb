module Github
  class Task
    class Destroy
      include Mandate

      initialize_with :issue_url

      def call
        Github::Task.where(issue_url:).destroy_all
      end
    end
  end
end
