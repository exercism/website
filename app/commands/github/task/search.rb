class Github::Task::Search
  include Mandate

  # Use class method rather than constant for
  # easier stubbing during testing
  def self.requests_per_page
    20
  end

  def initialize(actions: nil, knowledge: nil, areas: nil, sizes: nil, types: nil, repo_url: nil,
                 track_id: nil, order: nil, page: 1, sorted: true, paginated: true)
    @actions = actions
    @knowledge = knowledge
    @areas = areas
    @sizes = sizes
    @types = types
    @repo = repo_url
    @track_id = track_id
    @order = order
    @sorted = sorted
    @paginated = paginated
    @page = page
  end

  def call
    @tasks = Github::Task

    filter!
    sort! if sorted?
    paginate! if paginated?

    @tasks
  end

  private
  attr_reader :track_id, :actions, :knowledge, :areas, :sizes, :types,
    :repo, :order, :page, :sorted, :paginated, :tasks

  %i[sorted paginated].each do |attr|
    define_method("#{attr}?") { instance_variable_get("@#{attr}") }
  end

  def filter!
    %i[track_id repo actions knowledge areas sizes types].each do |filter|
      value = send(filter)
      @tasks = @tasks.where(filter.to_s.singularize.to_sym => value) if value.present?
    end
  end

  def sort!
    case order
    when "track"
      @tasks = @tasks.includes(:track).order('tracks.slug ASC')
    when "oldest"
      @tasks = @tasks.order(opened_at: :asc)
    else
      @tasks = @tasks.order(opened_at: :desc)
    end
  end

  def paginate!
    @tasks = @tasks.page(page).per(self.class.requests_per_page)
  end
end
