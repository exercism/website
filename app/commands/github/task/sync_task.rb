class Github::Task::SyncTask
  include Mandate

  initialize_with :issue

  def call
    return destroy_task! if closed? || claimed?
    return destroy_task! unless labeled?

    create_or_update_task!
  end

  private
  def closed?
    issue.status == :closed
  end

  def claimed?
    labels.any? { |label| label.name == Github::IssueLabel.for_type(:status, :claimed) }
  end

  def labeled?
    types = %i[action knowledge module size type]

    labels.any? do |label|
      types.any? do |type|
        label.of_type?(type)
      end
    end
  end

  def create_or_update_task!
    Github::Task::CreateOrUpdate.(
      issue.github_url,
      repo: issue.repo,
      title: issue.title,
      opened_at: issue.opened_at,
      opened_by_username: issue.opened_by_username,
      action: find_label_of_type(:action),
      knowledge: find_label_of_type(:knowledge),
      area: find_label_of_type(:module),
      size: find_label_of_type(:size),
      type: find_label_of_type(:type)
    )
  end

  def destroy_task!
    Github::Task::Destroy.(issue.github_url)
  end

  def find_label_of_type(type)
    labels.find { |l| l.of_type?(type) }&.value
  end

  memoize
  def labels
    issue.labels.to_a
  end
end
