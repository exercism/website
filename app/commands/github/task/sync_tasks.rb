class Github::Task::SyncTasks
  include Mandate

  def call
    issues.each do |issue|
      Github::Task::SyncTask.(issue)
    rescue StandardError => e
      Rails.logger.error "Error syncing task for issue #{issue.repo}/#{issue.number}: #{e}"
    end
  end

  private
  def issues
    Github::Issue.includes(:labels)
  end
end
