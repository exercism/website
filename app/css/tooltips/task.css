/* See also app/css/pages/contributing-tasks.css */
.c-task-tooltip {
    @apply shadow-lg rounded-8 py-20 px-24 bg-backgroundColorA;
    @apply relative z-tooltip;

    width: 540px;

    & section {
        @apply py-8 px-8;
        @apply flex items-center;

        &:first-child {
            @apply pt-0;
        }
        &:last-child {
            @apply pb-0;
        }

        & + section {
            @apply border-t-1 border-borderColor6;
        }

        & > .icon {
            @apply flex-shrink-0 grid place-items-center;
            @apply mr-24;
            width: 45px;

            .c-icon {
                @apply filter-textColor6;
            }

            & .action-icon {
                @apply border-none;
                .c-icon {
                    @apply filter-textColor6;
                }
            }
            & .knowledge-tag {
                @apply border-none p-0;
                & .c-icon {
                    @apply hidden;
                }
                & .dot {
                    height: 8px;
                    width: 8px;
                    margin-right: 3px;
                    margin-left: 0;
                }
            }
        }
        & h3 {
            @apply leading-150 font-medium;
            & strong {
                @apply font-semibold;
            }
        }
        & p {
            @apply leading-150;
        }
        &.summary {
            @apply pb-16;
            & .task-icon {
                @apply rounded-3 border-none;
                background: var(--backgroundColorMentorTestimonialsNew);
                color: var(--textColorMentorTestimonialsNew);
                @apply text-12 font-bold leading-150 font-mono uppercase;
                @apply py-4 px-8;
            }
            & h3 {
                @apply text-18 leading-150 font-medium;
            }
        }
    }
}
