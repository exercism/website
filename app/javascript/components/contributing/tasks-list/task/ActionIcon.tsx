// i18n-key-prefix: actionIcon
// i18n-namespace: components/contributing/tasks-list/task
import React from 'react'
import { GraphicalIcon, Icon } from '../../../common'
import { TaskAction } from '../../../types'
import { useAppTranslation } from '@/i18n/useAppTranslation'

export const ActionIcon = ({
  action,
}: {
  action?: TaskAction
}): JSX.Element => {
  const { t } = useAppTranslation()
  return (
    <div className="action-icon">
      {action ? (
        <Icon
          icon={`task-action-${action}`}
          alt={t('actionIcon.action', { action })}
        />
      ) : (
        <GraphicalIcon icon={`task-action`} />
      )}
    </div>
  )
}
