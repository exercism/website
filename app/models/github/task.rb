class Github::Task < ApplicationRecord
  disable_sti!

  belongs_to :track, optional: true

  enum action: {
    create: 0,
    fix: 1,
    improve: 2,
    proofread: 3,
    sync: 4
  }, _suffix: true

  enum knowledge: {
    none: 0,
    elementary: 1,
    intermediate: 2,
    advanced: 3
  }, _prefix: true

  enum area: {
    analyzer: 0,
    "concept-exercise": 1,
    concept: 2,
    generator: 3,
    "practice-exercise": 4,
    representer: 5,
    "test-runner": 6
  }, _suffix: true

  enum size: {
    tiny: 0,
    small: 1,
    medium: 2,
    large: 3,
    massive: 4
  }, _prefix: true

  enum type: {
    ci: 0,
    coding: 1,
    content: 2,
    docker: 3,
    docs: 4
  }, _suffix: true

  before_validation do
    self.track = Track.for_repo(repo) unless track
  end

  %i[action knowledge area size type].each do |type|
    define_method type do
      super()&.to_sym
    end
  end

  before_create do
    self.uuid = SecureRandom.compact_uuid if self.uuid.blank?
  end
end
