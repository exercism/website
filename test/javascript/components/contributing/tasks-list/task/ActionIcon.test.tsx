import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { ActionIcon } from '@/components/contributing/tasks-list/task/ActionIcon'

describe('ActionIcon tests', () => {
  test('ActionIcon snapshot', () => {
    const { asFragment } = render(<ActionIcon action="proofread" />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders a GraphicalIcon if action is not specified', () => {
    render(<ActionIcon />)

    const img = screen.queryByRole('img')
    expect(img).toBeInTheDocument()

    const altAttribute = img?.getAttribute('alt')
    expect(altAttribute).not.toContain('Action:')
  })

  test('renders an Icon with correct alt tag if action is specified', () => {
    render(<ActionIcon action="create" />)

    const img = screen.queryByRole('img')
    expect(img).toBeInTheDocument()

    const altAttribute = img?.getAttribute('alt')
    expect(altAttribute).toContain('Action: create')
  })
})
