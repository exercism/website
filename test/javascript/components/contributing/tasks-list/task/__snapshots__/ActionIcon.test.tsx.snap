// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionIcon tests ActionIcon snapshot 1`] = `
<DocumentFragment>
  <div
    class="action-icon"
  >
    <img
      alt="Action: proofread"
      class="c-icon"
      src="undefined/assets/undefined"
    />
  </div>
</DocumentFragment>
`;
